### PR TITLE
cylc gui: dbchooser: show host:port for suites + passphrase discovery logic refactor

### LINKED ISSUE #126: suite passphrase location?
To keep the `.cylc` directory tidy, the legal suite passphrase locations under `$HOME/.cylc/` should be moved down a level to `$HOME/.cylc/passphrases/`

See also Issue #124

### LINKED ISSUE #345: gcylc: File -> Open for switching to suites running on another host
From @benfitzpatrick - "t would be useful for us here if cylc gui --host=X detected suites running on X in the db-viewer part" [i.e. the dialog for opening registered suites].

### LINKED ISSUE #1763: remote tasks load wrong passphrase if suite name is registered remotely too
If I run a suite called SUITE on SUITE_HOST that submits a task to TASK_HOST, the remote task can load the wrong passphrase if TASK_HOST also has a suite registered as SUITE (with a passphrase in its suite definition directory).

### PR DESCRIPTION
**Pull request by matthewrmshin**
Close #126, close #345, close #1763.

(A better solution is to solve #1302, and have the *Switch To Another Suite* functionality launch `cylc gscan` instead?)

See commit log of 03e31792838ea24058e6cb775f33727849f4a603 for what's changed in the passphrase discovery logic refactor.

### CODE REVIEW

**Comment by hjoliver** [comment]
> have the Switch To Another Suite functionality launch cylc gscan instead?

I like this idea.

Relatedly, it we be good to get #1625 done - to make gscan as flexible as scan is now (see other user's suites, filter by user, name).

**Comment by matthewrmshin** [comment]
@hjoliver @benfitzpatrick please review. I think this change is good enough for solving the original issue. (But hopefully, we'll able to look at improving `cylc gscan` soon.)

(Note: the `scan_all` function is moved from `bin/cylc-scan` to `lib/cylc/network/port_scan.py`. The change is very small otherwise.)

**Comment by hjoliver** [comment]
@matthewrmshin - this doesn't seem to work as expected for me.  If I start up the GUI with `gcylc --host=OTHER` the File->Open dialog shows nothing, if host OTHER doesn't share the filesystem. If it does share the filesystem I see all my registered suites and can switch to a running suite on OTHER but it doesn't show up as running in the dialog.  So the only effect seems to be to get "localhost:PORT" in the dialog for running local suites??

**Comment by matthewrmshin** [comment]
@hjoliver That's interesting, because it appears to work for me. I have started a suite on a machine with a shared FS and a suite on a machine without shared FS - and the chooser works as long as I have got something like this in my `global.rc`.
```
[suite host scanning]
    hosts = localhost, host-with-share-fs, host-wo-share-fs
```

**Comment by hjoliver** [comment]
Ok, I'll have to look into this tomorrow...

**Comment by matthewrmshin** [comment]
Branch re-based.

**Comment by hjoliver** [comment]
Retested.  Now a remote suite (no shared filesystem) shows up in the chooser if I have a local suite registered with the same name, but clicking on it fails to connect due to "port file not found - suite not running?".   If I unregister the local suite, the remote one no longer appears in the chooser.

**Comment by matthewrmshin** [comment]
Interesting. I'll do some re-tests.

**Comment by hjoliver** [comment]
@matthewrmshin - I wonder if it is not worth fixing this, if we can just launch gscan instead?

**Comment by matthewrmshin** [comment]
> I wonder if it is not worth fixing this, if we can just launch gscan instead?

Yes, this may be the easiest.

**Comment by hjoliver** [comment]
@matthewrmshin - what do you want to do about this one?

**Comment by matthewrmshin** [comment]
I'll look at this one again after #1732 is merged, as this would conflict with it.

**Comment by matthewrmshin** [comment]
@hjoliver I can now see the problem you are having. I think I have fixed most of it apart from the passphrase requirement one. I'll see what can be done.

**Comment by matthewrmshin** [comment]
@hjoliver I think I have cracked this one with a refactor of the logic for loading/dumping passphrases. See log message of my latest commit for detail.

The most important change is the new functionality to load the passphrase of a remote suite via SSH (assuming that it is registered in the default `~/.cylc/REGDB/` of the remote host). Once retrieved, the passphrase will be saved locally, so it does not have to redo a SSH to load the passphrase.

The refactor should also allow us to close #126 as well.

Please take a look to see if you have any fundamental disagreement with the approach.

**Comment by hjoliver** [comment]
@matthewrmshin - looks good, I like it!   A couple of things I noticed testing this:

 * the File dialog shows running suites owned by others too, which the user is unlikely to have access to (but might do, so maybe that's ok...)
 * if a remote suite's passphrase gets changed after it is cached locally, the local copy will be wrong even though I technically have access to the correct one.  So is it worth a single attempted re-fetch of the remote passphrase on a "wrong passphrase" error?
 * if I attempt to connect to a remote suite owned by OTHER-USER that I do not have access to, the GUI doesn't respond sensibly - the status bar continually switches between "stopped" and "connected" (this probably isn't the fault of your branch).  Then, any subsequent attempt to connect to one of my own running suite MY-SUITE fails with "no object found by this name", 'OTHER-USER.MY-SUITE.suite-info'), or to one of my own stopped suites with "ERROR: security reasons - Incorrect suite passphrase?".  
 * If another user accidentally makes his passphrase readable, I automatically get full control of his suite now just by connecting to it, whereas before I'd have to look for the passphrase and steal it manually.   Is this desirable?  (actually I'm not bothered about this - users shouldn't make their secrets publicly available)
 * local and remote suites with same names cause some problems:
  * the local suite's passphrase will be loaded if I try to connect to the remote suite
  * the running remote suite will be shown in the dialog, but not the local one (running or not)

**Comment by matthewrmshin** [comment]
I have just realised that I have forgotten to do owner filtering after calling `scan_all` (which `cylc scan` does on the client side). I'll do this to address bullet point 1 (and hopefully reduces the likelihood of bullet point 3 and 4).

I'll see to bullet point 2 and 5.

**Comment by matthewrmshin** [comment]
The latest commit addresses bullet point 1 and 5, and should fix #1763.

**Comment by matthewrmshin** [comment]
The latest commits implement memory cache and retry-without-cache for passphrase loading. It is not tested yet, as I have run out of time for today.

**Comment by matthewrmshin** [comment]
Branch slightly squashed and re-based.

**Comment by matthewrmshin** [comment]
@hjoliver This should now be good for another look. I'll try and add some test battery for the passphrase caching logic tomorrow.

**Comment by hjoliver** [comment]
Looks good now.

**Comment by matthewrmshin** [comment]
Doc updated. Basic passphrase caching test added.

**Comment by matthewrmshin** [comment]
@hjoliver @benfitzpatrick please review.

**Comment by hjoliver** [comment]
@matthewrmshin - I'm getting some more problems with this branch (hadn't tested yesterday sorry, just looked over the code).  Trying to connect the GUI to a remote suite (which is visible in the dialog), I get `_get_proxy() got an unexpected keyword argument 'reset'`. I'm definitely running the head of this branch at both ends.   After that, attempting to open another stopped suite results in `('no object found by this name', 'oliverh.SUITE.suite-info')`, and a different  running remote suite `Choose a suite or cancel!`

File: doc/cug.tex [hunk 1]
```diff
@@ -876,10 +876,14 @@ \subsection{Suite Passphrases}
...
 A suite-specific passphrase file is automatically generated in the suite
 definition directory at registration time. It is loaded by the suite daemon at
-start-up and used to authenticate all client connections.  Clients on the suite
+start-up and used to authenticate all client connections. Clients on the suite
 host account automatically load the passphrase from the suite definition
-directory, but on other accounts, including task host accounts, it has to be
-installed manually.
+directory. The passphrase is also installed automatically on the suite run
+directory on hosts for running task jobs. On other accounts that do not share
+the same HOME directory with the suite host account and does not have
```
Comment (hjoliver) [inline doc/cug.tex]: "and do not"

File: doc/cug.tex [hunk 2]
```diff
@@ -5616,15 +5617,22 @@ \subsubsection{Full Control - Suite Passphrases}
 challenge-response scheme; they are never sent raw over the network.
 
 Clients on the suite host account automatically pick up the passphrase from the
-suite definition directory. On other accounts the passphrase has to be
-installed manually. This includes task host accounts unless the suite uses ssh
-messaging on that host (see~\ref{TaskComms}).  Allowed passphrase locations are:
+suite definition directory. On submission of the first task job on a host, the
+suite host will attempt to install the passphrase to the run directory to
+enable task jobs to connect to the suite. On other accounts that do not share
+the same HOME directory with the suite host account and does not have
```
Comment (hjoliver) [inline doc/cug.tex]: "do not"

**Comment by hjoliver** [comment]
I just tried again after deleting old passphrases and .pyc files (just in case).  This time I connected to the remote suite successfully, but then still got the problems listed above when trying to switch back to other suites.

File: lib/cylc/gui/dbchooser.py [hunk 3]
```diff
@@ -348,7 +362,7 @@ def __init__(self, parent, db, db_owner, tmpdir, pyro_timeout):
 
         cr = gtk.CellRendererText()
         tvc = gtk.TreeViewColumn(
-            'Port', cr, text=1, foreground=4, background=5)
+            'Server', cr, text=1, foreground=4, background=5)
```
Comment (hjoliver) [inline lib/cylc/gui/dbchooser.py]: Should be "Server:Port"?

**Comment by matthewrmshin** [comment]
@hjoliver all comments should be addressed.

**Comment by hjoliver** [comment]
Sorry - still got problems here!

1. start suite 'bar' on remote wrh-1:7767
  * (wrh-1 does not share a filesystem with my desktop box)
  * (no suite of that name registered locally either)
2. start gcylc on my desktop
3. use File menu to connect to bar on wrh-1
  * GOOD; title bar says *bar - wrh-1:7767*
4. use File to switch to local suite 'sig' (no suite of that name registered on wrh-1) - BAD
  * title bar says *sig - wrh-1:7767* 
  * attempt run 'sig', output to terminal is:
    ```
cylc run   --host=wrh-1.hpcf.niwa.co.nz --user=oliverh sig 
ERROR: Suite not found sig
ERROR: remote command terminated by signal 1
    ```
5.  switch back to 'bar' on wrh-1 - GOOD
6. switch to local stopped suite 'baz' (there IS a suite of that name reg'd on wrh-1) - BAD
     * error dialog: `_get_proxy() got an unexpected keyword argument 'reset'`
     * title bar says *baz - wrh-1:7767* ('bar' is listening on 7767, not 'baz')
     * can start the remote suite successfully, but the GUI is trying to connect to 'baz' (which reports lots of DENIED messages)
7. switch back to 'bar' - GOOD
8. switch again to 'baz', which now shows as running on 7771 - BAD
       * error dialog `Choose a suite or cancel!`

**Comment by hjoliver** [comment]
(double checked: I am running your latest version of this branch at both ends).

**Comment by matthewrmshin** [comment]
I'll have another look again today. (You can try running `cylc gui` in debug mode to get a traceback when it fails, just to give ourselves more information.)

**Comment by matthewrmshin** [comment]
@hjoliver The scenario described by your latest comment should be addressed by the latest commit.

**Comment by matthewrmshin** [comment]
Branch re-based.

**Comment by hjoliver** [comment]
This seems to be working now, but I can't fully test it from home (will do it tomorrow).

File: doc/cug.tex [hunk 4]
```diff
@@ -876,10 +876,14 @@ \subsection{Suite Passphrases}
 
 A suite-specific passphrase file is automatically generated in the suite
 definition directory at registration time. It is loaded by the suite daemon at
-start-up and used to authenticate all client connections.  Clients on the suite
+start-up and used to authenticate all client connections. Clients on the suite
 host account automatically load the passphrase from the suite definition
-directory, but on other accounts, including task host accounts, it has to be
-installed manually.
+directory. The passphrase is also installed automatically on the suite run
```
Comment (hjoliver) [inline doc/cug.tex]: *in* the suite run directory

File: lib/cylc/network/pyro_base.py [hunk 5]
```diff
@@ -166,70 +145,77 @@ def _set_uri(self):
...
-            self._get_proxy()
-            try:
-                self.pyro_proxy.signout()
-            except AttributeError:
-                # Back-compat for pre client reporting daemons <= 6.4.1.
-                pass
+            self._get_proxy().signout()
         except Exception:
             # Suite may have stopped before the client exits.
             pass
```
Comment (hjoliver) [inline lib/cylc/network/pyro_base.py]: Are we dropping all the backward compatibility in this module?

**Comment by hjoliver** [comment]
Tests as working now.

**Comment by hjoliver** [comment]
Test battery passes here too.  Review 1 done (I'm interested in the backward compat question above, but I don't particularly mind either way).

File: lib/cylc/network/pyro_base.py (hunk 5 above)
Comment (matthewrmshin) [inline lib/cylc/network/pyro_base.py]: Not intentionally, but I cannot see the reason why we need catch `AttributeError` separately here, now that `self._get_proxy()` returns `self.pyro_proxy`.

File: doc/cug.tex (hunk 4 above)
Comment (matthewrmshin) [inline doc/cug.tex]: Fixed.

**Comment by hjoliver** [comment]
Hang on a bit. This morning I did some more extensive testing, and I still get some problems.

I had one remote suite running, one or more local suites running (one had the same name as the local remote one, although I don't know if that matters), and played around with connecting, reconnecting, and starting and stopping suites at both ends (sometimes with `run --hold`, in the unlikely event that matters), deleting cached passphrases and restarting the GUI.

 Unfortunately I have not nailed down exactly how to trigger these problems, but nevertheless over 5-10 minutes of testing I got the following to occur multiple times:

 * failed to connect to the remote suite - "Choose a suite or cancel!" dialog on clicking "OK".  This happens more often than not.
 * completely blank File dialog - when this happens 'htop' shows several extras processes and/or threads (or both, I think) were spawned and do not go away, and then I can't exit gcylc, the process has to be killed.  
 * completely locked X server!  (I can get this to happen on both my desktop Centos box and my laptop Ubuntu VM).  I think the extra processes and/or threads exist here too.  To free up the system I have to go to a virtual terminal and kill the cylc-gui processes.

**Comment by matthewrmshin** [comment]
@hjoliver of your latest comment...

Bullet point 1 should now be fixed. It was a bit of silly logic.

Bullet point 2. I have seen this once or twice before, but after you have mentioned it, I was able to repeat this a lot if I open the dialog in quick successions. It was likely caused by the dialog calling `scan_all` every second. This should now be reduced to every minute. I have also reduce the sleep time of the main loop to 0.1 second, so it should be more responsive. I am no longer seeing this in my tests, but I may be wrong.

Bullet point 3. I have no idea what causes this, and I am so far unable to repeat.

Branch re-based.

**Comment by hjoliver** [comment]
Those problems all seem to be fixed now, thanks.   I think maybe the 3rd point was somehow caused by the second - I can't get it to happen again in any case.

**Comment by hjoliver** [comment]
Last point, relatively minor (I'm sure you're sick to death of this branch by now :grimacing:) - if I'm connected to a suite that shuts down, then it starts up again on a different port the GUI will not see it (even if I restart it again from the GUI!).  Can we unset the port number on "connection lost" and re-discover it next time a connection attempt is made?

**Comment by matthewrmshin** [comment]
Now reset port on stopped. Branch re-based.

**Comment by hjoliver** [comment]
Latest change tested.  I'm happy with this now, over to @benfitzpatrick

**Comment by matthewrmshin** [comment]
Branch re-based.

File: lib/cylc/gui/dbchooser.py [hunk 6]
```diff
@@ -202,52 +215,36 @@ def my_get_iter(item):
             iter = None
```
Comment (benfitzpatrick) [inline lib/cylc/gui/dbchooser.py]: Use `iter_` here if you are replacing it elsewhere?

File: lib/cylc/gui/dbchooser.py [hunk 7]
```diff
@@ -202,52 +215,36 @@ def my_get_iter(item):
...
-        results = scan(pyro_timeout=self.pyro_timeout)
-        choices = [(result[1]['name'], result[0]) for result in results]
-        choices.sort()
-        if choices != self.running_choices:
-            self.running_choices = choices
-            return True
-        else:
-            return False
+            gobject.idle_add(self.update)
+            sleep(0.1)
```
Comment (benfitzpatrick) [inline lib/cylc/gui/dbchooser.py]: Too fast?

File: lib/cylc/network/pyro_base.py [hunk 8]
```diff
@@ -88,31 +83,25 @@ def call_server_func(self, fname, *fargs):
...
-        try:
-            return func(*fargs)
-        except Pyro.errors.ConnectionDeniedError:
-            # Back compat for daemons <= 6.4.1: passphrase-only auth.
-            if cylc.flags.debug:
-                print >> sys.stderr, "Old daemon? - trying passphrases."
-            self.pyro_proxy = None
-            self._get_proxy_old()
-            func = getattr(self.pyro_proxy, fname)
+        items = [
```
Comment (benfitzpatrick) [inline lib/cylc/network/pyro_base.py]: I had hoped to nuke this back comp stuff soon, to be honest.

File: lib/cylc/gui/dbchooser.py [hunk 9]
```diff
@@ -63,65 +49,92 @@ def __init__(self, regd_treestore, db, filtr=None, pyro_timeout=None):
...
+        if self.next_scan_time is not None and self.next_scan_time > time():
+            return
+
+        # Scan for running suites
+        choices = []
+        for host, scan_result in scan_all(pyro_timeout=self.pyro_timeout):
+            try:
+                port, suite_identity = scan_result
+            except ValueError:
+                # Back-compat (<= 6.5.0 no title or state totals).
```
Comment (benfitzpatrick) [inline lib/cylc/gui/dbchooser.py]: Slightly misleading comment? state totals don't have to be public.

File: tests/authentication/10-remote-suite-passphrase-cache/suite.rc [hunk 10]
```diff
@@ -0,0 +1,14 @@
...
+        live mode suite timeout = PT1M
+[scheduling]
+    initial cycle point=1970
+    final cycle point=1970
+    [[dependencies]]
+        [[[P1Y]]]
+            graph=t1
+[runtime]
+    [[t1]]
+        script=printenv FOO
```
Comment (benfitzpatrick) [inline tests/authentication/10-remote-suite-passphrase-cache/suite.rc]: Won't fail if a user sets `FOO` in their `.profile`!

File: lib/cylc/gui/dbchooser.py (hunk 7 above)
Comment (matthewrmshin) [inline lib/cylc/gui/dbchooser.py]: I have found that 1 second was making the GUI a bit unresponsive, and could also have been the reason behind the strange blank screen + zombie behaviour. Reducing this to 0.1 makes the GUI more responsive, and I am no longer seeing the blank screen + zombie behaviour.

(See also the other logic, which delays calls to `scan_all` in 1-minute intervals.)

File: lib/cylc/network/pyro_base.py (hunk 8 above)
Comment (matthewrmshin) [inline lib/cylc/network/pyro_base.py]: Version 7, perhaps.

**Comment by benfitzpatrick** [comment]
Some comments, otherwise fine.

**Comment by matthewrmshin** [comment]
@benfitzpatrick Comments addressed and/or answered. I am now going to do some final tests.

**Comment by matthewrmshin** [comment]
Re-based and tested.

**Pull request merged**